agent/stats/engine_integ_test.go

```diff
@@ -220,89 +191,275 @@ func TestStatsEngineWithNewContainers(t *testing.T) {
+	time.Sleep(checkPointSleep)
+
+	// Verify the metrics of the container
+	metadata, taskMetrics, err := engine.GetInstanceMetrics()
+	assert.NoError(t, err, "error gettting instance metrics")
+	assert.NoError(t, validateMetricsMetadata(metadata), "error validating metadata")
+	assert.Len(t, taskMetrics, 1, "incorrect number of tasks")
+
 	taskMetric := taskMetrics[0]
-	if *taskMetric.TaskDefinitionFamily != taskDefinitionFamily {
-		t.Errorf("Excpected task definition family to be: %s, got: %s", taskDefinitionFamily, *taskMetric.TaskDefinitionFamily)
+	assert.Equal(t, aws.StringValue(taskMetric.TaskDefinitionFamily), "docker-container-health", "task definition family not expected")
+	assert.Equal(t, aws.StringValue(taskMetric.TaskDefinitionVersion), taskDefinitionVersion, "task definition family version not expected")
+	assert.NoError(t, validateContainerMetrics(taskMetric.ContainerMetrics, 1), "error validating container metrics")
+
+	// Verify the health metrics of container
+	healthMetadata, healthMetrics, err := engine.GetTaskHealthMetrics()
+	assert.NoError(t, err, "getting task health metrics failed")
+	require.Len(t, healthMetrics, 1)
+	assert.NoError(t, validateHealthMetricsMetadata(healthMetadata))
+	assert.Equal(t, aws.StringValue(healthMetrics[0].TaskArn), testTask.Arn, "task arn not expected")
+	assert.Equal(t, aws.StringValue(healthMetrics[0].TaskDefinitionFamily), "docker-container-health", "task definition family not expected")
+	assert.Equal(t, aws.StringValue(healthMetrics[0].TaskDefinitionVersion), taskDefinitionVersion, "task definition version not expected")
+	assert.NoError(t, validateTaskHealth(healthMetrics[0].Containers, 1))
+
+	err = client.StopContainer(container.ID, defaultDockerTimeoutSeconds)
+	assert.NoError(t, err, "stopping container failed")
+
+	err = engine.containerChangeEventStream.WriteToEventStream(ecsengine.DockerContainerChangeEvent{
+		Status: api.ContainerStopped,
+		DockerContainerMetadata: ecsengine.DockerContainerMetadata{
+			DockerID: container.ID,
+		},
+	})
+	assert.NoError(t, err, "write to container change event stream failed")
+
+	time.Sleep(waitForCleanupSleep)
+
+	// Should not contain any metrics after cleanup.
+	err = validateIdleContainerMetrics(engine)
+	assert.NoError(t, err, "validating idle metrics failed")
+
+	_, healthMetrics, err = engine.GetTaskHealthMetrics()
+	assert.NoError(t, err, "getting task health failed")
+	assert.Len(t, healthMetrics, 0, "no health metrics was expected")
+}
+
+func TestStatsEngineWithNewContainers(t *testing.T) {
+	// Create a new docker stats engine
+	engine := NewDockerStatsEngine(&cfg, dockerClient, eventStream("TestStatsEngineWithNewContainers"))
+	defer engine.removeAll()
+
+	container, err := createHealthContainer(client)
+	require.NoError(t, err, "creating container failed")
+	defer client.RemoveContainer(docker.RemoveContainerOptions{
+		ID:    container.ID,
+		Force: true,
+	})
+
+	// Wait for containers from previous tests to transition states.
```

Comment:
Why is this needed? Aren't we cleaning up containers properly at the end of the tests? This seems like a race condition that will lead to flaky tests because of failure on slower machines.

Reply:
This is still outstanding

Reply:
I'm not sure about that, it exists in the previous test. Looking at the code, it looks to me that the previous test should impact the current test, I'll remove it and see if it works.